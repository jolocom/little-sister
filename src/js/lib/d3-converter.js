import rdf from 'rdflib'
let FOAF = rdf.Namespace('http://xmlns.com/foaf/0.1/')
let DC = rdf.Namespace('http://purl.org/dc/terms/')
let RDF = rdf.Namespace('http://www.w3.org/1999/02/22-rdf-syntax-ns#')
import STYLES from 'styles/app.js'

// D3 Converter takes a node (a node in this context is an array of triples that
// describe an rdf document) and then, based on that returns an array where the
// triples are represented in a different format. This array can then be fed
// to D3 to draw a graph based on the data

class D3Converter {
  convertToD3(node, i, n) {
    // We need to know the index of the node and the total amount of nodes
    // in order to be able to calculate their initial position, so that they are
    // possitioned in a circle
    this.i = i + 1
    this.n = n

    let uri = node.uri

    let props = {
      uri: null,
      name:null,
      description:null,
      img:null,
      type:null,
      rank: null,
      x: null,
      y: null
    }

    // We create a rdf.graph() object, and populate it with the triples, this
    // allows us to then parse them using the rdflib's function
    // rdf.graph().statementsMatching()

    let g = rdf.graph()
    for (let i = 0; i < node.length; i++) {
      g.add(node[i].subject, node[i].predicate, node[i].object)
    }
    // Calculating the coordinates of the nodes so we can put them in a circle
    let angle = (2 * Math.PI) / this.n
    let halfwidth = STYLES.width / 2
    let halfheight = STYLES.height / 2

    props.x = Math.sin(angle * this.i) * halfwidth + halfwidth
    props.y =(Math.cos(angle * this.i) * halfwidth + halfheight)

    // Updating the attributes of the node object. The resulting object will have
    // all of it's props filled in, and will be ready to be rendered by D3
    // Note, if a triple is not present, it will be set to null.
    props.uri = uri

    // If the resource is a URI, it's value is stored next to the 'uri' key in the object
    // otherwise it's value is stored in the 'value' key of the object. We need to make
    // sure we are assigning the value regardless of where it's stored
    let name = g.statementsMatching(undefined, FOAF('name'), undefined)
    if (name.length > 0) props.name = name[0].object.value ? name[0].object.value : name[0].object.uri
    else props.name = null

    let description = g.statementsMatching(undefined, DC('description'), undefined)
    if (description.length > 0) props.description = description[0].object.value ? description[0].object.value : description[0].object.uri
    else props.description = null

    let type = g.statementsMatching(undefined, RDF('type'), undefined)
    if (type.length > 0) props.type = type[0].object.value ? type[0].object.value : type[0].object.uri
    else props.type = null

  //This labels the nodes as center and adjacent, a bit hacky, but it works allright
    if (i) props.rank = 'adjacent'
    if (!i) props.rank = 'center'

    return props
  }
}

export default D3Converter
