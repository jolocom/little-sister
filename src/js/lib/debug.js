import debug from 'debug'
import {flow, map} from 'lodash/fp'

debug.log = console.trace.bind(console)

// Prints an array of triples using console.table
console.rdftable = (array) => {
  flow(
      map((statement) => ({
        subject: statement.subject.uri || statement.subject.value,
        predicate: statement.predicate.uri.split("/").slice(-1)[0],
        object: statement.object.uri || statement.object.value
      })),
      console.table.bind(console))
  (array)
}

export default debug