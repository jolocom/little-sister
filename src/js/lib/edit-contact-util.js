import Immutable from 'immutable'

export const validateChanges = (state) => {
  const {newInformation, originalInformation: original} = state
    .get('information').toJS()
  const newFields = [...newInformation.emails, ...newInformation.phones]
  const oldFields = [...original.emails, ...original.phones]

  const isValid = newFields.every(e => (e.blank || e.valid || e.delete)) &&
    oldFields.every(e => e.delete || !e.update || (e.valid && !e.verified))

  return state.setIn(['showErrors'], !isValid)
}

const isBlank = (value, field) => {
  if (typeof value === 'object') {
    return !value.value || !value.value.trim()
  }
  return !value.trim()
}

const isValidValue = (value, field) => {
  switch (field) {
    case 'phones':
      return (/^([\d.]+)$/.test(value.value) ||
        /^\+([\d]+)$/.test(value.value))
    case 'emails':
      return /^([\w.]+)@([\w.]+)\.(\w+)/.test(value)
    default:
      return false
  }
}

const parseActionValueToObject = (value, field) => field === 'phones'
  ? value : {value}

const addTypeAttribute = (field) => field === 'phones'
  ? {type: 'personal'} : {}

export const setNewFieldValue = (state, {field, index, value}) => state.mergeIn(
  ['information', 'newInformation', field, index], {
    ...parseActionValueToObject(value, field),
    valid: isValidValue(value, field),
    blank: isBlank(value, field)
  })

export const mapAccountInformationToState = ({email, phone, addresses = []}) =>
  Immutable.fromJS({
    loading: false,
    showErrors: false,
    information: {
      newInformation: {
        emails: [email.length > 0 ? {delete: true} : {
          value: '',
          delete: false,
          blank: true,
          valid: true
        }],
        phones: [phone.length > 0 ? {delete: true} : {
          value: '',
          type: 'personal',
          blank: true,
          delete: false,
          valid: true
        }],
        addresses: [addresses.length > 0 ? {delete: true} : {
          streetWithNumber: {value: '', valid: true},
          zip: {value: '', valid: true},
          city: {value: '', valid: true},
          state: {value: '', valid: true},
          country: {value: '', valid: true},
          delete: false,
          blank: true,
          valid: true
        }]
      },
      originalInformation: {
        emails: email.map(({address, verified, id}) => ({
          value: address,
          verified,
          id,
          delete: false,
          update: false,
          valid: true
        })),
        phones: phone.map(({number, verified, id}) => ({
          value: number,
          // TODO: phone type needs to be delivered from the backend
          type: 'personal',
          verified,
          id,
          delete: false,
          update: false,
          valid: true
        })),
        addresses: addresses.map(({streetWithNumber, zip, city, state, country}) => ({ // eslint-disable-line max-len
          streetWithNumber: {...streetWithNumber, valid: true},
          zip: {...zip, valid: true},
          city: {...city, valid: true},
          state: {...state, valid: true},
          country: {...country, valid: true},
          delete: false,
          blank: true,
          valid: true
        }))
      }
    }
  })

const canAddNewField = (state, field, index) => {
  if (index === 0) { return true }
  if (!state
    .getIn(['information', 'newInformation', field, index - 1, 'delete'])) {
    return !state
      .getIn(['information', 'newInformation', field, index - 1, 'blank'])
  }
  return canAddNewField(state, field, index - 1)
}

export const addNewField = (state, {field, index}) => {
  if (field === 'addresses') {
    return state.mergeIn(['information', 'newInformation', field, index], {
      streetWithNumber: {value: '', valid: true},
      zip: {value: '', valid: true},
      city: {value: '', valid: true},
      state: {value: '', valid: true},
      country: {value: '', valid: true},
      delete: false,
      blank: true,
      valid: true
    })
  }
  if (canAddNewField(state, field, index)) {
    return state.mergeIn(['information', 'newInformation', field, index], {
      ...addTypeAttribute(field), value: '', verified: false, valid: false,
      delete: false, blank: true
    })
  }
  return state
}

const canUpdate = (state, field, index) => !state.getIn(
  ['information', 'originalInformation', field, index, 'verified']) &&
  state.getIn(['information', 'originalInformation', field]).toJS().length > 0

export const updateOriginalValue = (state, {field, value, index}) => {
  if (canUpdate(state, field, index)) {
    return state.mergeIn(['information', 'originalInformation', field, index], {
      ...parseActionValueToObject(value, field),
      valid: isValidValue(value, field),
      delete: value === '',
      update: true
    })
  }
  return state
}
const collectChages =
  (state, {remove, update, set}, key, webId, attributeType) => [].concat(
    state.originalInformation[key].map(
      e => e.delete ? remove(webId, attributeType, e.id)
      : (e.update && e.valid && !e.verified)
      ? update(webId, attributeType, e.id, e.value)
      : null
    ), state.newInformation[key].map(
      e => (e.delete || e.blank || !e.valid)
      ? null : set(webId, e.value)
  ))

export const submitChanges = (backend, services, state, webId) => {
  let solidAgent = backend.solid
  const emailOperations = {
    set: solidAgent.setEmail.bind(solidAgent),
    remove: solidAgent.deleteEntry.bind(solidAgent),
    update: solidAgent.updateEntry.bind(solidAgent)
  }
  const phoneOperations = {
    set: solidAgent.setPhone.bind(solidAgent),
    remove: solidAgent.deleteEntry.bind(solidAgent),
    update: solidAgent.updateEntry.bind(solidAgent)
  }
  let promises = [].concat(
    collectChages(state, emailOperations, 'emails', webId, 'email'),
    collectChages(state, phoneOperations, 'phones', webId, 'phone'))
  return Promise.all(promises)
}
