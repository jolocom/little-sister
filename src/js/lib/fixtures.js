export default {}

export let bankUri = 'https://localhost:8443'

export let contacts = [{
  username: 'eelco',
  name: 'Eelco Wiersma',
  email: 'hi@eelcowiersma.nl'
}, {
  username: 'joachim',
  name: 'Joachim Lohkamp',
  email: 'joachim@jolocom.com'
}, {
  username: 'justas',
  name: 'Justas Azna',
  email: 'justas@jolocom.com'
}, {
  username: 'christian',
  name: 'Christian Hildebrand',
  email: 'christian@jolocom.com'
}, {
  username: 'anna',
  name: 'Anna Blume',
  email: 'anna@jolocom.com'
}]

export let conversations = [{
  type: 'user',
  username: 'justas',
  items: [{
    author: {
      username: 'justas',
      name: 'Justas Azna'
    },
    content: 'This is completely fucked up!',
    date: new Date()
  }]
}, {
  type: 'user',
  username: 'joachim',
  items: [{
    author: {
      username: 'joachim',
      name: 'Joachim Lohkamp'
    },
    content: 'How are you?',
    date: new Date()
  }]
}]
