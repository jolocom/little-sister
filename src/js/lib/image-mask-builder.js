export default class ImageMaskBuilder {
  constructor() {
    this.paths = []
    this.currentPath = null
  }

  startNewPath() {
    this.currentPath = []
    this.paths.push(this.currentPath)
  }

  endPath() {
    this.currentPath = null
  }

  addPoint(x, y) {
    // if (x === 0 && y === 0) {
    //   return
    // }

    this.currentPath.push([x, y])
  }
}
