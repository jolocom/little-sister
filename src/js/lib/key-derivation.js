/* @summary - Generates a keypair based on provided entropy
 *
 * @param {Mnemonic} seed - a Bitcore-Mnemonic generated from hashed entropy.
 * @returns {HDPrivateKey} - an HDPrivateKey instance containing a keypair.
*/

export function deriveMasterKeyPair(seed) {
  return seed.toHDPrivateKey()
}

export function derivePrivateChildKeyPair(masterPrivateKeyPair, path) {
  return masterPrivateKeyPair.derive(path)
}

/* @summary - Generate a generic signing key according to TODO
 *
 * @param {HDPrivateKey} masterPrivateKeyPair - the master keypair used to
 * derive the child key.
 *
 * @returns {HDPrivateKey} - the derived generic signing key.
 *
*/

export function deriveGenericSigningKeyPair(masterPrivateKeyPair) {
  return derivePrivateChildKeyPair(masterPrivateKeyPair, 'm/73\'/0\'/0\'')
}

/* @summary - Generate an Ethereum keypair according to goo.gl/sr5dvy
 *
 * @param {HDPrivateKey} masterPrivateKeyPair - the master keypair used to
 * derive the child key.
 *
 * @returns {HDPrivateKey} - the derived Ethereum keypair.
 *
*/

export function deriveEthereumKeyPair(masterPrivateKeyPair) {
  return derivePrivateChildKeyPair(masterPrivateKeyPair, 'm/44\'/60\'/0\'/0/0')
}
