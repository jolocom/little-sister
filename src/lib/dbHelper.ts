import * as squel from 'squel'

interface Field {
  name: string
  type: string
  options: string[]
}

export interface TableOptions {
  name: string
  fields: Field[]
}

export interface PersonaAttributes {
  did: string
  controllingKey: string
}

export interface AssembledQuery {
  text: string
  values?: string[]
}

const PersonaTable = {
  name: 'Personas',
  fields: [{
    name: 'did',
    type: 'VARCHAR(75)',
    options: ['PRIMARY KEY', 'NOT NULL']
  }, {
    name: 'controllingKey',
    type: 'VARCHAR(110)',
    options: ['NOT NULL', 'UNIQUE']
  }, {
    name: 'FOREIGN KEY(controllingKey)',
    type: 'REFERENCES Keys(encryptedWif)',
    options: []
  }]
}

// TODO Unique on entr src + algo + path
const KeysTable = {
  name: 'Keys',
  fields: [{
    name: 'encryptedWif',
    type: 'VARCHAR(110)',
    options: ['PRIMARY KEY', 'NOT NULL']
  }, {
    name: 'path',
    type: 'TEXT',
    options: ['NOT NULL']
  }, {
    name: 'entropySource',
    type: 'VARCHAR(100)',
    options: ['NOT NULL']
  }, {
    name: 'keyType',
    type: 'TEXT',
    options: ['NOT NULL']
  }, {
    name: 'FOREIGN KEY(entropySource)',
    type: 'REFERENCES MasterKeys(encryptedEntropy)',
    options: []
  }]
}

const MasterKeysTable = {
  name: 'MasterKeys',
  fields: [{
    name: 'encryptedEntropy',
    type: 'VARCHAR(100)',
    options: ['PRIMARY KEY', 'NOT NULL']
  }, {
    name: 'timestamp',
    type: 'INTEGER',
    options: ['NOT NULL']
  }]
}

const VerifiableCredentialsTable = {
  name: 'VerifiableCredentials',
  fields: [{
    name: 'context',
    type: 'TEXT',
    options: ['NOT NULL']
  }, {
    name: 'id',
    type: 'VARCHAR(50)',
    options: ['PRIMARY KEY', 'NOT NULL']
  }, {
    name: 'type',
    type: 'TEXT',
    options: ['NOT NULL']
  }, {
    name: 'name',
    type: 'VARCHAR(20)',
    options: ['NOT NULL']
  }, {
    name: 'issuer',
    type: 'VARCHAR(75)',
    options: ['NOT NULL']
  }, {
    name: 'issued',
    type: 'INTEGER',
    options: ['NOT NULL']
  }, {
    name: 'expiry',
    type: 'INTEGER',
    options: []
  }, {
    name: 'subject',
    type: 'VARCHAR(75)',
    options: ['NOT NULL']
  }, {
    name: 'FOREIGN KEY(subject)',
    type: 'REFERENCES Personas(did)',
    options: []
  }]
}

const ClaimsTable = {
  name: 'Claims',
  fields: [{
    name: 'credentialId',
    type: 'VARCHAR(50)',
    options: ['NOT NULL']
  }, {
    name: 'propertyName',
    type: 'VARCHAR(50)',
    options: ['NOT NULL']
  }, {
    name: 'encryptedValue',
    type: 'TEXT',
    options: ['NOT NULL']
  }, {
    name: 'PRIMARY KEY (credentialId, propertyName)',
    type: '',
    options: []
  }, {
    name: 'UNIQUE (credentialId, propertyName)',
    type: '',
    options: []
  }, {
    name: 'FOREIGN KEY(credentialId)',
    type: 'REFERENCES VerifiableCredentials(id)',
    options: []
  }]
}

const SignatureTable = {
  name: 'Signatures',
  fields: [{
    name: 'signatureType',
    type: 'VARCHAR(20)',
    options: ['NOT NULL']
  }, {
    name: 'signatures',
    type: 'TEXT',
    options: ['NOT NULL']
  }, {
    name: 'credentialId',
    type: 'VARCHAR(50)',
    options: ['PRIMARY KEY', 'NOT NULL']
  }, {
    name: 'FOREIGN KEY(credentialId)',
    type: 'REFERENCES VerifiableCredentials(id)',
    options: []
 }]
}

export const dbHelper = {
  defaultTables: [
    PersonaTable,
    KeysTable,
    MasterKeysTable,
    ClaimsTable,
    VerifiableCredentialsTable,
    SignatureTable
  ],

  createTableQuery: (options: TableOptions) : AssembledQuery => {
    const { name, fields } = options
    const st = `CREATE TABLE IF NOT EXISTS ${name}`

    const fieldSt = fields.map(f => {
      const { name, type, options } = f
      const fieldAttributes = options.join(' ')
      return `${name} ${type} ${fieldAttributes}`
    }).join(', ')
    return { text: `${st} (${fieldSt})` }
  },

  addPersonaQuery: (args: PersonaAttributes) : AssembledQuery => {
    const { did, controllingKey } = args
    return squel.insert()
      .into('Personas')
      .setFields({
        did,
        controllingKey
      })
      .toParam()
  },

  addDerivedKeyQuery: (args: DerivedKeyAttributes) : AssembledQuery => {
    const { encryptedWif, path, entropySource, keyType } = args
    return squel.insert()
      .into('Keys')
      .setFields({
        encryptedWif,
        path,
        entropySource,
        keyType
      })
      .toParam()
  },

  addMasterKeyQuery: (encryptedEntropy: string) : AssembledQuery => {
    return squel.insert()
      .into('MasterKeys')
      .setFields({
        encryptedEntropy,
        timestamp: Date.now()
      })
      .toParam()
  },

  // TODO START WILD WEST
  // TODO Should this take a claim object?
  addVerifiableClaimQuery: (args: any) => {
    const testClaim = {
      "@context": [
        "https://w3id.org/identity/v1",
        "https://w3id.org/security/v1"
      ],
      "id": "http://example.gov/credentials/3732",
      "type": ["Credential", "PassportCredential"],
      "name": "Passport",
      "issuer": "https://example.gov",
      "issued": "2010-01-01",
      "claim": {
        "id": "did:example:ebfeb1f712ebc6f1c276e12ec21",
        "name": "Alice Bobman",
        "birthDate": "1985-12-14",
        "gender": "female",
        "nationality": {
          "name": "United States"
        },
        "address": {
          "type": "PostalAddress",
          "addressStreet": "372 Sumter Lane",
          "addressLocality": "Blackrock",
          "addressRegion": "Nevada",
          "postalCode": "23784",
          "addressCountry": "US"
        },
        "passport": {
          "type": "Passport",
          "name": "United States Passport",
          "documentId": "123-45-6789",
          "issuer": "https://example.gov",
          "issued": "2010-01-07T01:02:03Z",
          "expires": "2020-01-07T01:02:03Z"
        }
      },
      "signature": {
        "type": "LinkedDataSignature2015",
        "created": "2016-06-21T03:40:19Z",
        "creator": "https://example.com/jdoe/keys/1",
        "domain": "json-ld.org",
        "nonce": "783b4dfa",
        "signatureValue": "Rxj7Kb/tDbGHFAs6ddHjVLsHDiNyYzxs2MPmNG8G47oS06N8i0Dis5mUePIzII4+p/ewcOTjvH7aJxnKEePCO9IrlqaHnO1TfmTut2rvXxE5JNzur0qoNq2yXl+TqUWmDXoHZF+jQ7gCsmYqTWhhsG5ufo9oyqDMzPoCb9ibsNk="
      }
    }

    const signature = dbHelper.extractSignature(testClaim)
    const claim = dbHelper.extractClaim(testClaim)
    console.log(claim)
    console.log(signature)
  },

  // TODO FILTER
  extractClaim: (claim) => {
    return {
      id: claim.claim.id,
      claims: claim.claim
    }
  },

  extractSignature: (claim) => {
    const result = {
      type: '',
      signature: ''
    }

    const supportedKeys = ['signature', 'signatureChain']
    supportedKeys.some(k => {
      if (claim[k]) {
        result.type = k
        result.signature = claim[k]
      }
      return true
    })

    return result
  },

  // TODO END WILD WEST
  getPersonasQuery: () : AssembledQuery => {
    return squel.select()
      .from('Personas')
      .toParam()
  }
}
