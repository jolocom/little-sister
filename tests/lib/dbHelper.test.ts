import * as MockDate from 'mockdate'
import { dbHelper } from 'src/lib/dbHelper'
import data from './testData/storageTestData'

describe.only('dbHelper', () => {
  it('should correctly assemble table creation query', () => {
    const { personasTableInfo, keysTableInfo } = data
    expect(dbHelper.createTableQuery(personasTableInfo)).toMatchSnapshot()
    expect(dbHelper.createTableQuery(keysTableInfo)).toMatchSnapshot()
  }),

  it('should correctly assemble derived key addition query', () => {
    const mockArgs = {
      encryptedWif: 'mockWif',
      path: 'm/73\'/0\'/0',
      entropySource: 'mockEncryptedWif',
      keyType: 'ECDSA secp256k1'
    }

    expect(dbHelper.addDerivedKeyQuery(mockArgs)).toMatchSnapshot()
  }),

  it('should correctly assemble master key addition query', () => {
    MockDate.set(765000000000, 0)
    expect(dbHelper.addMasterKeyQuery('mockEntropy')).toMatchSnapshot()
    MockDate.reset()
  })

  it('should correctly assemble persona addition query', () => {
    expect(dbHelper.addPersonaQuery({
      did: 'mockDid',
      controllingKey: 'mockEncryptedWif'
    })).toMatchSnapshot()
  })

  it('should correctly assemble persona retrieval query', () => {
    expect(dbHelper.getPersonasQuery()).toMatchSnapshot()
  })
})

