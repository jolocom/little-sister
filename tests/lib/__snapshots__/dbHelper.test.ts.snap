// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbHelper should correctly assemble derived key addition query 1`] = `
Object {
  "text": "INSERT INTO Keys (encryptedWif, path, entropySource, keyType) VALUES (?, ?, ?, ?)",
  "values": Array [
    "mockWif",
    "m/73'/0'/0",
    "mockEncryptedWif",
    "ECDSA secp256k1",
  ],
}
`;

exports[`dbHelper should correctly assemble master key addition query 1`] = `
Object {
  "text": "INSERT INTO MasterKeys (encryptedEntropy, timestamp) VALUES (?, ?)",
  "values": Array [
    "mockEntropy",
    946681200000,
  ],
}
`;

exports[`dbHelper should correctly assemble persona addition query 1`] = `
Object {
  "text": "INSERT INTO Personas (did, controllingKey) VALUES (?, ?)",
  "values": Array [
    "mockDid",
    "mockEncryptedWif",
  ],
}
`;

exports[`dbHelper should correctly assemble table creation query 1`] = `
Object {
  "text": "CREATE TABLE IF NOT EXISTS Personas (did VARCHAR(20) NOT NULL PRIMARY KEY, FOREIGN KEY(controllingKey) REFERENCES Keys(id) )",
}
`;

exports[`dbHelper should correctly assemble table creation query 2`] = `
Object {
  "text": "CREATE TABLE IF NOT EXISTS Keys (id INTEGER PRIMARY KEY NOT NULL UNIQUE, FOREIGN KEY(entropySource) REFERENCES MasterKeys(id) )",
}
`;
